doc/design/docker-cp.md

```diff
@@ -0,0 +1,144 @@
+
+OR 
+
+* Error - error response from the portlayer
+
+```
+
+### Personality design
+
+The personality has two endpoints that are called depending on the to/from situation above. Each of those endpoints will call their respective portlayer counterparts. There will be several checks during this process. 
+
+Error codes expected by docker : 
+
+```
+200 
+
+400
+
+403 - in the event of a readonly filesystem
+
+404
+
+500
+
+```
+
+Because we need to prevent a poweroff(unless forced) of a container during a copy, we will pass the ID directly to the portlayer rather than fetching a handle. 
+
+Docker does not do much in the way of string checking as far as the path is concerned. This is likely not a big concern for the personality as the actual copy commmand will fail if a bad path string is provided. 
+
+The personality code should be fairly straight forward, the goal is to call the portlayer for the copy while passing the tar stream either to the portlayer or from it. The best way to do this is to hook the writiers together using a PipeReader and then shepherding the data in the direction it needs to go. 
+
+## Portlayer Design
+
+ There will be two situations to be concerned with in the portlayer. When the container is *on* and when the container is *off*.
+ 
+ 
+### Off Container Scenario
+ 
+ This is the case where guest tools is not needed since the guest in off. This means that all action will need to occur on the VCH. There are several concerns and many of them are listed in this ticket [here](https://github.com/vmware/vic/issues/4105)
+ 
+ The most concern surrounding this piece of work will be surrounding VCH restart and confirming that the disk is still left in a good state if a copy was occuring mid restart. Additionally we need to block a container from starting when we have the r/w layer mounted. While that container is blocked from being started we might look at the possibility of returning a 500 error for a start invocation on a blocked container. We must also block other copy operations until the current one is successful, one major reason is because we cannot do any kind of sanity checking on available space with concurrent writes to the same disk. 
+ 
+ Additionally, it must be ensured that regardless of error or success the disk must be detached from the VCH. We must ensure that this also occurs when we boot up the VCH in the event that the disk is attached to the VCH and is hard shutdown. If the VCH does not already have a disk detach functionality at boot time we should be adding this. This is an ideal time to invoke calls to `fsck` and `journal` in order to recover from mid copy shutdowns. Since the VCH is meant to be stateless(and the portlayer) we should aim to jettison and disk that is attached to the VCH at boot time if we do write this portion of the portlayer(We should even make it a separate ticket as testing and ensuring quality on this could take time.)
```

Comment:
There is already code (and I think a test) for ensuring that disks are detached on VCH boot if present.
Unsure that `journal` needs to be run explicitly for ext4 - the journal should be replayed on mount regardless (unless `noload`) is specified. Likewise fsck is of marginal value for a journaled filesystem.
There is this: https://github.com/vmware/vic/issues/4872#issuecomment-305608290 which suggests we may be exposed in some manner, but we'd need to determine what sort of writes could lead us to that circumstance (O_DIRECT maybe?)